src/runtime_src/core/pcie/driver/linux/xocl/subdev/flash.c

```diff
@@ -18,13 +19,1109 @@
+		}
+
+		cnt += thislen;
+	}
+
+	mutex_unlock(&flash->io_lock);
+
+	vfree(page);
+	if (ret)
+		return ret;
+
+	*off += n;
+	return n;
+}
+
+/*
+ * Write a page. Perform read-modify-write as needed.
+ * @cnt contains actual bytes copied from user on successful return.
+ */
+static int flash_page_rmw(struct xocl_flash *flash,
+	const char __user *ubuf, u8 *kbuf, loff_t off, size_t *cnt)
+{
+	loff_t thisoff = FLASH_PAGE_ALIGN(off);
+	size_t front = FLASH_PAGE_OFFSET(off);
+	size_t mid = min(*cnt, FLASH_PAGE_SIZE - front);
+	size_t last = FLASH_PAGE_SIZE - front - mid;
+	u8 *thiskbuf = kbuf;
+	int ret;
+
+	if (front) {
+		ret = flash_buf_rdwr(flash, thiskbuf, thisoff, front, false);
+		if (ret)
+			return ret;
+	}
+	thisoff += front;
+	thiskbuf += front;
+	if (copy_from_user(thiskbuf, ubuf, mid) != 0)
+		return -EFAULT;
+	*cnt = mid;
+	thisoff += mid;
+	thiskbuf += mid;
+	if (last) {
+		ret = flash_buf_rdwr(flash, thiskbuf, thisoff, last, false);
+		if (ret)
+			return ret;
+	}
+
+	ret = flash_page_erase(flash, FLASH_PAGE_ALIGN(off), FLASH_PAGE_SIZE);
+	if (ret == 0) {
+		ret = flash_buf_rdwr(flash, kbuf, FLASH_PAGE_ALIGN(off),
+			FLASH_PAGE_SIZE, true);
+	}
+	return ret;
+}
+
+static inline size_t flash_get_page_io_size(loff_t off, size_t sz)
+{
+	if (IS_ALIGNED(off, FLASH_HUGE_PAGE_SIZE) &&
+		sz >= FLASH_HUGE_PAGE_SIZE)
+		return FLASH_HUGE_PAGE_SIZE;
+	if (IS_ALIGNED(off, FLASH_LARGE_PAGE_SIZE) &&
+		sz >= FLASH_LARGE_PAGE_SIZE)
+		return FLASH_LARGE_PAGE_SIZE;
+	if (IS_ALIGNED(off, FLASH_PAGE_SIZE) &&
+		sz >= FLASH_PAGE_SIZE)
+		return FLASH_PAGE_SIZE;
+
+	return 0; // can't do full page IO
+}
+
+/*
+ * Try to erase and write full (large/huge) page.
+ * @cnt contains actual bytes copied from user on successful return.
+ * Needs to fallback to RMW, if not possible.
+ */
+static int flash_page_wr(struct xocl_flash *flash,
+	const char __user *ubuf, u8 *kbuf, loff_t off, size_t *cnt)
+{
+	int ret;
+	size_t thislen = flash_get_page_io_size(off, *cnt);
+
+	if (thislen == 0)
+		return -EOPNOTSUPP;
+
+	*cnt = thislen;
+
+	if (copy_from_user(kbuf, ubuf, thislen) != 0)
+		return -EFAULT;
+
+	ret = flash_page_erase(flash, off, thislen);
+	if (ret == 0)
+		ret = flash_buf_rdwr(flash, kbuf, off, thislen, true);
+	return ret;
+}
+
+/*
+ * Write to flash memory page by page from user buf.
+ */
+static ssize_t
+flash_write(struct file *file, const char __user *buf, size_t n, loff_t *off)
+{
+	struct xocl_flash *flash = file->private_data;
+	u8 *page = NULL;
+	size_t cnt = 0;
+	int ret = 0;
+
+	if (n == 0 || *off + n >= flash->flash_size)
+		return -EINVAL;
+
+	page = vmalloc(FLASH_HUGE_PAGE_SIZE);
+	if (page == NULL)
+		return -ENOMEM;
+
+	mutex_lock(&flash->io_lock);
+
+	if (!flash_wait_till_ready(flash))
+		ret = -EINVAL;
+
+	while (ret == 0 && cnt < n) {
+		loff_t thisoff = *off + cnt;
+		const char *thisbuf = buf + cnt;
+		size_t thislen = n - cnt;
+
+		/* Try write full page. */
+		ret = flash_page_wr(flash, thisbuf, page, thisoff, &thislen);
```

Comment:
Why is there a read,modify,write sequence for flash here? Images stored in flash have to be aligned to 4KB boundaries so you should always be able to do a 4KB erase and not need to store contents of what was in flash at that offset.

Reply:
this is a generic standard char device interface. user may not do page aligned write at all.